px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp

```diff
@@ -0,0 +1,99 @@
+/****************************************************************************
+ * Copyright (c) 2023 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_ros2/common/context.hpp>
+
+using namespace std::chrono_literals; // NOLINT
+
+namespace px4_ros2
+{
+/** \ingroup odometry
+ *  @{
+ */
+
+/**
+ * Provides a subscription to arbitrary ROS topics.
+ */
+template<typename RosMessageType>
+class Subscription
+{
+public:
+  Subscription(Context & context, const std::string & topic)
+  : _node(context.node())
+  {
+    const std::string namespaced_topic = context.topicNamespacePrefix() + topic;
+    _subscription = _node.create_subscription<RosMessageType>(
+      namespaced_topic, rclcpp::QoS(1).best_effort(),
+      [this](const typename RosMessageType::UniquePtr msg) {
+        _last = *msg;
+        _last_message_time = _node.get_clock()->now();
+        for (const auto & callback : _callbacks) {
+          callback(_last);
+        }
+      });
+  }
+
+  /**
+   * @brief Add a callback to execute when receiving a new message.
+   *
+   * @param callback the callback function
+   */
+  void onUpdate(std::function<void(RosMessageType)> callback)
```

Comment:
You could add a `using UpdateCallback = std::function<void(const RosMessageType&)>`

```suggestion
  void onUpdate(const std::function<void(const RosMessageType&)>& callback)
```